src/hotspot/share/gc/g1/g1EvacuationFailureRegions.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, Huawei Technologies Co. Ltd. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "g1EvacuationFailureRegions.hpp"
+#include "gc/g1/g1CollectedHeap.hpp"
+#include "gc/g1/g1CollectedHeap.inline.hpp"
+#include "gc/g1/heapRegion.hpp"
+#include "runtime/atomic.hpp"
+
+
+G1EvacuationFailureRegions::G1EvacuationFailureRegions() :
+  _regions_failed_evacuation(mtGC) {
+}
+
+G1EvacuationFailureRegions::~G1EvacuationFailureRegions() {
+  FREE_C_HEAP_ARRAY(uint, _evac_failure_regions);
+}
+
+void G1EvacuationFailureRegions::initialize() {
+  Atomic::store(&_evac_failure_regions_cur_length, 0u);
+  _max_regions = G1CollectedHeap::heap()->max_reserved_regions();
```

Comment:
Please pass in `max_reserved_regions()` - this makes the dependency on it during initialization obvious.